concordium-consensus/src/Concordium/GlobalState/Account.hs

```diff
@@ -249,13 +253,26 @@ data AccountMerkleHashInputs (av :: AccountVersion) where
         { -- | Hash of the persisting account data.
           amhi2PersistingAccountDataHash :: !PersistingAccountDataHash,
           -- | Hash of the account stake.
-          amhi2AccountStakeHash :: !(AccountStakeHash av),
+          amhi2AccountStakeHash :: !(AccountStakeHash 'AccountV2),
```

Comment:
It is still used, I think. Each constructor gives a different type :)

Reply:
But couldn't we drop the parametrization in line 251 then, becaues `av` doesn't appear in the constructors anymore.

Reply:
It "appears" in the constructors in the sense that the return type of the constructors depend on it and are different (the type `AccountMerkleHashInputs 'AccountV2` and the type `AccountMerkleHashInputs 'AccountV3`), so I guess it depends on whether we want the constructors to have different return types :)

Reply:
I still don't get it, `av` is literally not used anymore, independent whether you have different constructors for different versions. I think I'd try to drop the `av` parameter and see whether ghc is still able to compile it.

Reply:
Technically, the `av` in the `data AccountMerkleHashInputs (av :: AccountVersion) where` does not scope the constructor defintion, so in fact the old defintion allowed an `AccountStakeHash` of any version -- which is why this is fixed. The parameter is required for `AccountMerkleHashInputs` to have kind `AccountVersion -> *`, which is what we want. We're using a GADT here so that we can explicitly distinguish the type of `AccountMerkleHashInputs` on the `AccountVersion`, with `AccountMerkleHashInputsV2` only used for `AccountV2` and `AccountMerkleHashInputsV3` only used for `AccountV3`. If we remove the `av`, then we get rid of the type parameter and it's no longer a GADT -- and the type system gives us a little less of a guarantee.

Reply:
I guess I fail to see the additional guarantee by the type system. I think now you could put anything in the `av` parameter and choose any of the constructors. Maybe it's still nice to track the intention which `AccountVersion` we're constructing the `AccountMerkleHashInputs` for though.